src/block_cache/block_tier.c

```diff
@@ -0,0 +1,137 @@
+/*-
+ * Copyright (c) 2014-present MongoDB, Inc.
+ * Copyright (c) 2008-2014 WiredTiger, Inc.
+ *	All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+#include "wt_internal.h"
+
+/*
+ * __wt_blkcache_tiered_open --
+ *     Open a tiered object.
+ */
+int
+__wt_blkcache_tiered_open(
+  WT_SESSION_IMPL *session, const char *uri, uint32_t objectid, WT_BLOCK **blockp)
+{
+    WT_BLOCK *block;
+    WT_BUCKET_STORAGE *bstorage;
+    WT_CONFIG_ITEM pfx;
+    WT_DECL_ITEM(tmp);
+    WT_DECL_RET;
+    WT_TIERED *tiered;
+    const char *cfg[2], *object_name, *object_uri, *object_val;
+    bool exist, local_only, readonly;
+
+    *blockp = NULL;
+
+    tiered = (WT_TIERED *)session->dhandle;
+    object_uri = object_val = NULL;
+
+    WT_ASSERT(session, objectid <= tiered->current_id);
+    WT_ASSERT(session, uri == NULL || WT_PREFIX_MATCH(uri, "tiered:"));
+    WT_ASSERT(session, (uri == NULL && objectid != 0) || (uri != NULL && objectid == 0));
+
+    /*
+     * First look for the local file. This will be the fastest access and we retain recent objects
+     * in the local database for awhile. If we're passed a name to open, then by definition it's a
+     * local file.
```

Comment:
The recent discussion of coordination (or lack of it) around removing files we've retained in this manner makes me wonder if this is something we want as part of the long term design.  We know that eventually we are going to want a general purpose way to cache data from object storage on a local file system.  Should we take some time to decide how that is going to work and whether/how it pulls data into that cache after we flush a file to object storage before we worry about addressing the file removal race?

Just dropping this note here since this comment triggered the thought and all of the interested folks are reviewing the PR.  This is clearly outside the scope of this change.

Reply:
There are a variety of options here -- for example, you could cache data from object storage, in DRAM, NVRAM, the OS buffer cache, or local files, and it can be at either the file-level or individual block level. It's a big discussion.

Reply:
I'd love to be in the loop with respect to caching discussions.

Reply:
@keithbostic : in this context I'm specifically thinking about caching in local storage.  But I agree we should be conscious of the other options because inevitably we'll want some (all?) of them in the future.

@fedorova I'd love to have your input when we get to caching.